proposals/native-lock.md

```diff
@@ -0,0 +1,121 @@
+>    {
+>        class Lock
+>        {
+>            public Scope EnterLockScope();
+>    
+>            public ref struct Scope
+>            {
+>                public void Dispose();
+>            }
+>        }
+>    }
+>    ```
+> 2. where `x` is an expression of a *reference_type*, is precisely equivalent to: [...]
+
+Additionally, new warnings are added to implicit reference conversions ([§10.2.8](https://github.com/dotnet/csharpstandard/blob/9af5bdaa7af535f34fbb7923e5406e01db8489f7/standard/conversions.md#1028-implicit-reference-conversions))
+when upcasting the `System.Threading.Lock` type:
+
+> The implicit reference conversions are:
+>
+> - From any *reference_type* to `object` and `dynamic`.
+>   - **A warning is reported when the *reference_type* is known to be `System.Threading.Lock`.**
+> - From any *class_type* `S` to any *class_type* `T`, provided `S` is derived from `T`.
+>   - **A warning is reported when `S` is known to be `System.Threading.Lock`.**
+> - From any *class_type* `S` to any *interface_type* `T`, provided `S` implements `T`.
+>   - **A warning is reported when `S` is known to be `System.Threading.Lock`.**
+> - [...]
+
+```cs
+object l = new System.Threading.Lock(); // warning
+lock (l) { } // monitor-based locking is used here
+```
+
+Note that this warning occurs even for equivalent explicit conversions.
+To escape out of the warning and force use of monitor-based locking, one can use
+- the usual warning suppression means (`#pragma warning disable`),
+- `Monitor` APIs directly,
+- indirect casting like `object AsObject<T>(T l) => (object)l;`.
+
+## Alternatives
+[alternatives]: #alternatives
+
+- Support a general pattern that other types can also use to interact with the `lock` keyword.
+  This is a future work that might be implemented when `ref struct`s can participate in generics.
+  Discussed in [LDM 2023-12-04](https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-12-04.md#lock-statement-pattern).
+
+- To avoid ambiguity between the existing monitor-based locking and the new `Lock` (or pattern in the future), we could:
+  - Introduce a new syntax instead of reusing the existing `lock` statement.
+  - Require the new lock types to be `struct`s (since the existing `lock` disallows value types).
+    There could be problems with default constructors and copying if the structs have lazy initialization.
+
+- The codegen could be hardened against thread aborts (which are themselves obsoleted).
+
+- We could warn also when `Lock` is passed as a type parameter, because locking on a type parameter always uses monitor-based locking:
+
+  ```cs
+  M(new Lock()); // could warn here
+
+  void M<T>(T x) // (specifying `where T : Lock` makes no difference)
+  {
+      lock (x) { } // because this uses Monitor
+  }
+  ```
+
+  However, that would cause warnings when storing `Lock`s in a list which is undesirable:
+
+  ```cs
+  List<Lock> list = new();
+  list.Add(new Lock()); // would warn here
```

Comment:
>list.Add(new Lock()); // would warn here [](http://example.com/codeflow?start=2&length=40)

Why not warn here? Is storing a `Lock` instance in a `List<Lock>`, or passing the instance to a generic method, a likely scenario? (What about calling `GetHashCode()`?) It seems like the warnings could be suppressed easily for these cases.

Reply:
> Is storing a `Lock` instance in a `List<Lock>`, or passing the instance to a generic method, a likely scenario?

I think yes, this was brought by the runtime folks at the LDM as a reason why to not warn for generics IIRC.